### PR TITLE
add migration code which changes schema of CloudService in v1.11.0

### PR DESCRIPTION
**Pull request by stat-kwon**
### Category
- [x] New feature
- [ ] Bug fix
- [ ] Improvement
- [ ] Refactor
- [ ] etc

### Description
* add migration code which change schema of CloudService in v1.11.0
* add `hashlib` in requirements.txt

### CODE REVIEW

File: src/migration/v1_11_0.py [hunk 1]
```diff
@@ -40,11 +43,93 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+                    tag_provider = str(tag.get('provider', 'custom'))
+
+                    hash_value = string_to_hash(tag_key)
+
+                    if new_tags.get(tag_provider):
+                        new_tags[tag_provider].update({hash_value: {'key': tag_key, 'value': tag_value}})
+                    else:
+                        new_tags[tag_provider] = {hash_value: {'key': tag_key, 'value': tag_value}}
+
+                    if tag_type == 'PROVIDER' or 'MANAGED':
```
Comment (whdalsrnt) [inline src/migration/v1_11_0.py:91]: change condition
`if tag_type != 'CUSTOM':`

File: src/migration/v1_11_0.py [hunk 2]
```diff
@@ -40,11 +43,93 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+
+                for tag in tags:
+                    tag_key = str(tag['key'])
+                    tag_value = str(tag['value'])
+                    tag_type = str(tag['type'])
+                    tag_provider = str(tag.get('provider', 'custom'))
+
+                    hash_value = string_to_hash(tag_key)
+
+                    if new_tags.get(tag_provider):
```
Comment (whdalsrnt) [inline src/migration/v1_11_0.py:86]: Change tags code without provider
`new_tags[tag_provider] = new_tags.get(tag_provider, {})`

File: src/migration/v1_11_0.py [hunk 3]
```diff
@@ -40,11 +43,93 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+                        new_tags[tag_provider].update({hash_value: {'key': tag_key, 'value': tag_value}})
+                    else:
+                        new_tags[tag_provider] = {hash_value: {'key': tag_key, 'value': tag_value}}
+
+                    if tag_type == 'PROVIDER' or 'MANAGED':
+                        if new_tag_keys.get(tag_provider):
```
Comment (whdalsrnt) [inline src/migration/v1_11_0.py:92]: Change tags code without provider
`new_tag_keys[tag_provider] = new_tag_keys.get(tag_provider, {})`

File: src/migration/v1_11_0.py [hunk 4]
```diff
@@ -40,11 +43,93 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+                    else:
+                        if new_tag_keys.get('custom'):
+                            new_tag_keys['custom'].append(tag_key)
+                        else:
+                            new_tag_keys['custom'] = [tag_key]
+
+                update_fields['$set'].update({'tags': new_tags})
+                update_fields['$set'].update({'tag_keys': new_tag_keys})
+
+            elif not tags and isinstance(tags, list):
```
Comment (whdalsrnt) [inline src/migration/v1_11_0.py:105]: change condition
`if isinstance(tags, list):`

**Review by whdalsrnt — CHANGES_REQUESTED** [review]

File: src/migration/v1_11_0.py (hunk 1 above)
Comment (stat-kwon) [reply src/migration/v1_11_0.py:91]: have reflected. thank you

File: src/migration/v1_11_0.py (hunk 3 above)
Comment (stat-kwon) [reply src/migration/v1_11_0.py]: have reflected. thank you

File: src/migration/v1_11_0.py [hunk 5]
```diff
@@ -40,11 +43,87 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+                update_fields['$set'].update({'tag_keys': new_tag_keys})
+
+            elif isinstance(tags, list):
+                update_fields['$set'].update({'tags': {}})
+
+            if metadata and provider not in metadata:
+                for plugin_id in metadata:
+                    update_fields['$set'].update({'metadata': {provider: metadata[plugin_id]}})
+
+                new_metadata = {provider: metadata}
```
Comment (whdalsrnt) [inline src/migration/v1_11_0.py:106]: Please delete this code.

File: src/migration/v1_11_0.py [hunk 6]
```diff
@@ -40,11 +43,87 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+                        new_tag_keys['custom'].append(tag_key)
+
+                update_fields['$set'].update({'tags': new_tags})
+                update_fields['$set'].update({'tag_keys': new_tag_keys})
+
+            elif isinstance(tags, list):
+                update_fields['$set'].update({'tags': {}})
+
+            if metadata and provider not in metadata:
+                for plugin_id in metadata:
```
Comment (whdalsrnt) [inline src/migration/v1_11_0.py:103]: change a syntax

```
new_metadata = {}
for plugin_id in metadata:
  new_metadata = {provider: metadata[plugin_id]}

update_fields['$set'].update(new_metadata)
```

File: src/migration/v1_11_0.py [hunk 7]
```diff
@@ -40,11 +43,87 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+                    update_fields['$set'].update({'metadata': {provider: metadata[plugin_id]}})
+
+                new_metadata = {provider: metadata}
+                update_fields['$set'].update({'metadata': new_metadata})
+
+            elif not metadata and provider in metadata:
```
Comment (whdalsrnt) [inline src/migration/v1_11_0.py:109]: Please delete this code.

**Review by whdalsrnt — CHANGES_REQUESTED** [review]

File: src/migration/v1_11_0.py [hunk 8]
```diff
@@ -40,11 +43,87 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+                new_tags = {}
+                new_tag_keys = {}
+
+                for tag in tags:
+                    tag_key = str(tag['key'])
+                    tag_value = str(tag['value'])
+                    tag_type = str(tag['type'])
+                    tag_provider = str(tag.get('provider', 'custom'))
+
+                    hash_value = string_to_hash(tag_key)
```
[the remark this answers is not in the document]
Comment (stat-kwon) [reply src/migration/v1_11_0.py:84]: have reflected. thank you

File: src/migration/v1_11_0.py [hunk 9]
```diff
@@ -40,11 +43,87 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+
+                    new_tags[tag_provider] = new_tags.get(tag_provider, {})
+                    new_tags[tag_provider].update({hash_value: {'key': tag_key, 'value': tag_value}})
+
+                    if tag_type != 'CUSTOM':
```
[the remark this answers is not in the document]
Comment (stat-kwon) [reply src/migration/v1_11_0.py:89]: have reflected. thank you

File: src/migration/v1_11_0.py [hunk 10]
```diff
@@ -40,11 +43,87 @@ def monitoring_alert_refactor_alert_number_by_domain_id(mongo_client: MongoCusto
...
+                        new_tag_keys[tag_provider] = new_tag_keys.get(tag_provider, [])
+                        new_tag_keys[tag_provider].append(tag_key)
+                    else:
+                        new_tag_keys['custom'] = new_tag_keys.get(tag_provider, [])
+                        new_tag_keys['custom'].append(tag_key)
+
+                update_fields['$set'].update({'tags': new_tags})
+                update_fields['$set'].update({'tag_keys': new_tag_keys})
```
[the remark this answers is not in the document]
Comment (stat-kwon) [reply src/migration/v1_11_0.py:97]: have reflected. thank you

File: src/migration/v1_11_0.py (hunk 5 above)
Comment (stat-kwon) [reply src/migration/v1_11_0.py:106]: have reflected. thank you

File: src/migration/v1_11_0.py (hunk 6 above)
Comment (stat-kwon) [reply src/migration/v1_11_0.py:103]: have reflected. thank you

File: src/migration/v1_11_0.py (hunk 7 above)
Comment (stat-kwon) [reply src/migration/v1_11_0.py:109]: have reflected. thank you

**Review by whdalsrnt — APPROVED** [review]

**Pull request merged**